File: libnd4j/tests_cpu/layers_tests/ArrayOptionsTests.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// Created by raver119 on 13.01.2018.
//
#include <array/ArrayOptions.hXX>
#include <array/NDArray.h>

#include "testlayers.h"

using namespace sd;

class ArrayOptionsTests : public NDArrayTests {
 public:
  LongType shape[8] = {2, 5, 5, 5, 1, 0, 1, 99};
};

TEST_F(ArrayOptionsTests, TestShape_Basic_0) {
  shape[5] = 1;

  ASSERT_TRUE(ArrayOptions::isNewFormat(shape));
  ASSERT_FALSE(ArrayOptions::isSparseArray(shape));
}

TEST_F(ArrayOptionsTests, TestShape_Basic_1) {
  shape[5] = 2;

  ASSERT_TRUE(ArrayOptions::isNewFormat(shape));
  ASSERT_TRUE(ArrayOptions::isSparseArray(shape));
}

TEST_F(ArrayOptionsTests, TestShape_Basic_2) {
  shape[5] = 258;

  ASSERT_TRUE(ArrayOptions::isNewFormat(shape));

  ASSERT_TRUE(ArrayOptions::isSparseArray(shape));
  ASSERT_EQ(SpaceType::CONTINUOUS, ArrayOptions::spaceType(shape));
}

TEST_F(ArrayOptionsTests, TestShape_Basic_3) {
  ASSERT_EQ(0, shape::extra(shape));

  ASSERT_EQ(SpaceType::CONTINUOUS, ArrayOptions::spaceType(shape));
}

TEST_F(ArrayOptionsTests, TestShape_Basic_4) {
  ArrayOptions::setPropertyBits(shape, {ARRAY_HALF, ARRAY_QUANTIZED});

  auto dtype = ArrayOptions::dataType(shape);

  ASSERT_FALSE(ArrayOptions::isSparseArray(shape));
  ASSERT_TRUE(sd::DataType::HALF == ArrayOptions::dataType(shape));
  ASSERT_EQ(sd::ArrayType::DENSE, ArrayOptions::arrayType(shape));
  ASSERT_EQ(sd::SpaceType::QUANTIZED, ArrayOptions::spaceType(shape));
}

TEST_F(ArrayOptionsTests, TestShape_Basic_5) {
  ArrayOptions::setPropertyBits(shape, {ARRAY_SPARSE, ARRAY_INT, ARRAY_CSC});

  ASSERT_TRUE(ArrayOptions::isSparseArray(shape));
  ASSERT_TRUE(sd::DataType::INT32 == ArrayOptions::dataType(shape));
  ASSERT_EQ(sd::SparseType::CSC, ArrayOptions::sparseType(shape));
}

TEST_F(ArrayOptionsTests, TestShape_Basic_6) {
  ArrayOptions::setPropertyBits(shape, {ARRAY_EMPTY, ARRAY_INT, ARRAY_CSC});

  ASSERT_EQ(sd::ArrayType::EMPTY, ArrayOptions::arrayType(shape));
}

TEST_F(ArrayOptionsTests, TestShape_Basic_7) {
  ArrayOptions::setDataType(shape, FLOAT32);
  ArrayOptions::setDataType(shape, FLOAT32);

  ASSERT_EQ(sd::DataType::FLOAT32, ArrayOptions::dataType(shape));
}

TEST_F(ArrayOptionsTests, TestShape_Basic_8) {
  ArrayOptions::setDataType(shape, DOUBLE);
  ArrayOptions::setDataType(shape, FLOAT32);

  ASSERT_EQ(sd::DataType::FLOAT32, ArrayOptions::dataType(shape));
}

TEST_F(ArrayOptionsTests, TestShape_Basic_9) {
  ArrayOptions::setDataType(shape, FLOAT32);
  ArrayOptions::setDataType(shape, DOUBLE);

  ASSERT_EQ(sd::DataType::DOUBLE, ArrayOptions::dataType(shape));
}

File: nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/linalg/options/ArrayOptionsTests.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  *  See the NOTICE file distributed with this work for additional
 *  *  information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.options;

import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

import org.nd4j.common.tests.tags.TagNames;
import org.nd4j.linalg.BaseNd4jTestWithBackends;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.shape.options.ArrayOptionsHelper;
import org.nd4j.linalg.api.shape.options.ArrayType;
import org.nd4j.linalg.factory.Nd4jBackend;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;

@Slf4j
@Tag(TagNames.JAVA_ONLY)
public class ArrayOptionsTests extends BaseNd4jTestWithBackends {

    @ParameterizedTest
    @MethodSource("org.nd4j.linalg.BaseNd4jTestWithBackends#configs")
    public void testArrayType_0(Nd4jBackend backend) {
        long[]  shapeInfo = new long[]{2, 2, 2, 2, 1, 0, 1, 99};
        assertEquals(ArrayType.DENSE, ArrayOptionsHelper.arrayType(shapeInfo));
    }

    @ParameterizedTest
    @MethodSource("org.nd4j.linalg.BaseNd4jTestWithBackends#configs")
    public void testArrayType_1(Nd4jBackend backend) {
        long[]  shapeInfo = new long[]{2, 2, 2, 2, 1, 0, 1, 99};
        ArrayOptionsHelper.setOptionBit(shapeInfo, ArrayType.EMPTY);

        assertEquals(ArrayType.EMPTY, ArrayOptionsHelper.arrayType(shapeInfo));
    }

    @ParameterizedTest
    @MethodSource("org.nd4j.linalg.BaseNd4jTestWithBackends#configs")
    public void testArrayType_2(Nd4jBackend backend) {
        long[]  shapeInfo = new long[]{2, 2, 2, 2, 1, 0, 1, 99};
        ArrayOptionsHelper.setOptionBit(shapeInfo, ArrayType.SPARSE);

        assertEquals(ArrayType.SPARSE, ArrayOptionsHelper.arrayType(shapeInfo));
    }

    @ParameterizedTest
    @MethodSource("org.nd4j.linalg.BaseNd4jTestWithBackends#configs")
    public void testArrayType_3(Nd4jBackend backend) {
        long[]  shapeInfo = new long[]{2, 2, 2, 2, 1, 0, 1, 99};
        ArrayOptionsHelper.setOptionBit(shapeInfo, ArrayType.COMPRESSED);

        assertEquals(ArrayType.COMPRESSED, ArrayOptionsHelper.arrayType(shapeInfo));
    }

    @ParameterizedTest
    @MethodSource("org.nd4j.linalg.BaseNd4jTestWithBackends#configs")
    public void testDataTypesToFromLong(Nd4jBackend backend) {

        for(DataType dt : DataType.values()){
            if(dt == DataType.UNKNOWN)
                continue;
            String s = dt.toString();
            long l = 0;
            l = ArrayOptionsHelper.setOptionBit(l, dt);
            assertNotEquals(0, l,s);
            DataType dt2 = ArrayOptionsHelper.dataType(l);
            assertEquals(dt, dt2,s);
        }

    }

    @Override
    public char ordering() {
        return 'c';
    }
}
